rdflib/graph.py

```diff
@@ -670,35 +670,108 @@ def set(self, triple):
         self.add((subject, predicate, object_))
         return self
 
-    def subjects(self, predicate=None, object=None) -> Iterable[Node]:
-        """A generator of subjects with the given predicate and object"""
-        for s, p, o in self.triples((None, predicate, object)):
-            yield s
-
-    def predicates(self, subject=None, object=None) -> Iterable[Node]:
-        """A generator of predicates with the given subject and object"""
-        for s, p, o in self.triples((subject, None, object)):
-            yield p
-
-    def objects(self, subject=None, predicate=None) -> Iterable[Node]:
-        """A generator of objects with the given subject and predicate"""
-        for s, p, o in self.triples((subject, predicate, None)):
-            yield o
-
-    def subject_predicates(self, object=None):
+    def subjects(self, predicate=None, object=None, unique=False) -> Iterable[Node]:
+        """A generator of unique subjects with the given predicate and object"""
+        if not unique:
+            for s, p, o in self.triples((None, predicate, object)):
+                yield s
+        else:
+            subs = set()
+            for s, p, o in self.triples((None, predicate, object)):
+                if s not in subs:
+                    yield s
+                    try:
+                        subs.add(s)
+                    except MemoryError as e:
+                        print(f"{e}. Consider not setting parameter 'unique' to True")
```

Comment:
I think it is better to use [`logging.warning`](https://docs.python.org/3/library/logging.html#logging.warning)/[`Logger.warning`](https://docs.python.org/3/library/logging.html#logging.Logger.warning) instead of print - that way users can exercise some control over how this message is handled.

```suggestion
                        logger.error(f"{e}. Consider not setting parameter 'unique' to True")
```

Though most likely at the point where MemoryError occurs things will be too late.